### ISSUE TITLE
Feature Request: Configurable Fruit Press

### ISSUE THREAD

**Issue by PitchBright** [GrowthcraftCE/Growthcraft-1.7#61]
This idea came to me in using the Thirst Mod. It has a Drink Brewer block that you can stick something into, and get a drink in return. The inputs and outputs (drinks) are configurable, and the user essentially creates their own drinks via config file. That includes drink naming, drink colour, thirst replenish, saturation, poisoning chance, etc. It's quite ingenious.

The Fruit Press in GC is awesome. It would be great to be able to configure it to accept whatever item we would like… and then output whatever we would like. Perhaps the pressed items would make "juice" (or a fluid). If these things could be configured (and created) in the same way the Thirst Mod allows its users to literally create their own items via config, that would be wicked.

For reference, here's the source code for Thirst Mod: https://github.com/thetorine/thirstmod

Ideally this is done in such a way that the resulting items/fluids/drink from GC's Fruit Press can be used as Drinks in the Thirst Mod.

**Comment by Ar97x** [comment]
Perhaps we can create another machine, something like a juicer or a blender, and letting people choose whatever it want to make, just like you mentioned. Maybe we can have Tarun's help on making the drinks from GC work on Thirst Mod, I've tried to do it before (not so hard) but I failed. This is something I'm looking forward, since I play with both mods for a long time. This could be our first interaction to another mod, as it looks easier than the others cogitated before. 
@alatyami 
@PitchBright

**Comment by Yulife** [comment]
If you guys want to natively add Minetweaker support another complete configurationability via configs would be double the work. In my opinion, just stick with MT support, it's configureable to most extend.

**Comment by PitchBright** [comment]
@Ar97x But the fruit press looks and works so amazing. It's got that turn-of-the-centry "Pioneer" feel to it. The Drink Brewer serves as a machine/juicer/blender just fine… it's just that it's not nearly as cool as the Fruit Press. That was my angle with it.

@Yulife I'm not sure what you mean. I use both Minetweaker and Thirstmod, and I'm not making the connect with what you're saying. Can you clarify, please?

**Comment by PitchBright** [comment]
@Ar97x I actually think I could get you the Thirst Mod, if you wanted it. I've been talking to the author for a few years, and I might be able to talk him into letting you carry the torch. He's on-and-off with the mod, and doesn't/can't really support it, for the demand it garners. I think the lack of support is it's biggest downfall, otherwise it would probably be standard and "must-have" by most SMP players by now. People just give up trying to make it work, and they don't get answers to their questions. It's buggy, and it takes forever to get fixes.

**Comment by Ar97x** [comment]
I have worked with him some time ago (portuguese translations are mine). My only fear about taking the mod is that I don't know shit about programming. I can do a decent work responding questions and stuff, updating forums and making it look cooler, but when someone comes with a technical error I would be completely lost. I love that mod and I do know that all it needs is real support, and I don't care doing this for free, but I would need somebody to carry the torch with me, to help me with coding and new features, all this stuff. About the Fruit Press, I got your point, you like the way it works and think it would fit with the purpose. For me is perfectly fine, if anyone can do it would be amazing.

**Comment by Yulife** [comment]
@PitchBright iirc Minetweaker support was planned so why would you need config options to define recipes for the Fruit Press?

**Comment by PitchBright** [comment]
@Yulife I was thinking because it's not a matter of simply connecting a pre-existing input to a pre-existing output. The Thirst Mod provides you with the final product in a glass bottle, but the Fruit Press produces a liquid (an in-between stage) and you would need to be able to configure the press to create your own liquids, where none exists. From there, you would click on the liquid-filled Fruit Press with your empty bottle, and yield a bottle of juice (from the Thirst Mod for example).

**Comment by Yulife** [comment]
I'm sorry but your whole Thirst mod stuff confuses me. haha

**Comment by IceDragon200** [comment]
If someone is willing to take up the task for this, you could make a companion mod which takes a json/config file and injects the recipes into the CellarRegistry, no point in shoving this into the main growthcraft code base, its hectic as is :P (have you seen localization and integration code D8)

**Comment by IceDragon200** [comment]
@PitchBright [User Pressing Recipes](https://github.com/Alatyami/Growthcraft-1.7/wiki/User-API-%282.4.x%29#pressing-recipes)

**Comment by PitchBright** [comment]
Looks like christmas came early!

I'll test it up and confirm it works… and then close this one out.

I may hit a couple questions in the process. The first that pops to mind before I even begin is, do the fluids need to be legit fluids? Or does the press create a sort of pseudo-fluid that can't be poured out of a bucket?

**Comment by PitchBright** [comment]
I haven't tried anything yet, but I think to add to that, my next questions are going to be, using HarvestCraft as an example. 

How does pressing HarvestCraft Lemons in the Fruit Press, outputting Lemon Juice, become a HarvestCraft Lemon Juice drink when I tried to scoop a bottle of the juice from the Press? Speaking as a non-coder, how do those things connect?

And (depending on your answer to the first question about legit v pseudo juice)… is the resulting liquid from the press, something that can be stored in a Barrel (using Lemon as an example)?

**Comment by IceDragon200** [comment]
@PitchBright If HarvestCraft registered a fluid container (the bottle) for the lemon juice (the fluid) it will pick it up, otherwise you'll just have a press full of lemon juice.

tl;dr
HarvestCraft must register:
- A fluid (the lemon juice)
- A fluid container (a bottle of lemon juice)

If those are both true, then yes you'll have a nice bottle of lemon juice, otherwise you'll just have lemon juice fluid in the fruit press.

Growthcraft Barrels, Jars, etc.. will accept ANY fluid (go ahead and store Buildcraft Oil in it, if that's your thing ;x)

**Comment by PitchBright** [comment]
Awesome, thanks for the clarification!

**Comment by PitchBright** [comment]
This seems to be working beautifully!!!

Sorry it took so long to confirm that. Since Harvestcraft didn't have any fluids, I had to learn Java, and add my own. That only took a couple hundred hours :P 

(Which reminds me I wanted to ask you how you got your blockfluids to have their colour, in CreativeTabs… I couldn't seem to solve that one. If you can point to it in the code, I'd be grateful. EDIT: Also how you were able to make it so that bottles don't fill up when clicking on a fluid's source block)

Anyway, I have one last request on this, which I hope isn't a bother. Since folks that will want to customize their presses, will probably be doing it for realism… would it be possible to get the hardcoded Apple and Grapes, moved out of their module configs… and over to the Cellar Config… so that we can change out the bonemeal as the Pomace ourselves (for the purpose of configuring it to something like seeds, for example)?

**Comment by PitchBright** [comment]
Hit a little thing… I'm not sure if I'm entering it wrong or not… but I can't seem to make Pumpkins make a Fluid and pomace Pumpkin Seeds. It's not accepting the pumpkin. Because it's an itemblock maybe?

**Comment by PitchBright** [comment]
Ditto for Sugarcane/Reed into fluid and Sugar.

**Comment by IceDragon200** [comment]
@PitchBright We handle fluid colors here:

[Fluid Block Color](https://github.com/GrowthcraftCE/Growthcraft-1.7/blob/development/src/java/growthcraft/cellar/common/block/BlockFluidBooze.java#L88-L92)

In addition, we use an extended Fluid class: [Booze](https://github.com/GrowthcraftCE/Growthcraft-1.7/blob/development/src/java/growthcraft/api/cellar/booze/Booze.java#L38-L48) 
Which has a `setColor` method, and overrides the `getColor` method

Finally you'll need a special `ItemBlock` for your fluids
[ItemBlockFluidBooze](https://github.com/GrowthcraftCE/Growthcraft-1.7/blob/development/src/java/growthcraft/cellar/common/item/ItemBlockFluidBooze.java)

You'll need to override the `getColorFromItemStack` and return your fluid's color

**Comment by IceDragon200** [comment]
@PitchBright Here is a [Registry Dump](https://gist.github.com/IceDragon200/99388594cf92c9fd23e8) from Growthcraft's development. This may help you finding the item names for pumpkin and reeds

**Pumpkin** (Item)

``` json
{
  "item": {
    "mod_id": "minecraft",
    "name": "pumpkin",
    "amount": 1,
    "meta": 0
  }
}
```

**Pumpkin Seeds**

``` json
{
  "item": {
    "mod_id": "minecraft",
    "name": "pumpkin_seeds",
    "amount": 1,
    "meta": 0
  }
}
```

**Reed** (Item)

``` json
{
  "item": {
    "mod_id": "minecraft",
    "name": "reeds",
    "amount": 1,
    "meta": 0
  }
}
```

**Comment by PitchBright** [comment]
@IceDragon200 Yeah that's what I had in my config. Either I'm poochin' it somehow because I'm too tired… or something's up with 'em because their not "items" (maybe?).

@everyone Can anybody test those to confirm I'm not goin' crazy?

**Comment by IceDragon200** [comment]
@PitchBright You can enable debugging for the cellar module and watch the logs?

`config/growthcraft/cellar.conf`

```
debug {
    # Should Growthcraft log all its activity for debugging purposes?; Default : false
    B:"Enable Debugging"=true
}
```

Note that each module has its own debug option, so you can set them to debug individual modules.

Anything related to Brewing, Pressing, Fermenting, Distlling (when we get to it) is apart of the Cellar module (which is also why its JSON configs are under `config/growthcraft/cellar/*.json`)

Look out for these logs: (ignore the logger.error part, your interested in the strings)

``` java
logger.error("NULL RECIPE");
logger.error("No pressing item for recipe %s", recipe);
logger.error("Invalid item for recipe %s", recipe);
logger.error("No result fluid for recipe %s", recipe);
logger.error("Invalid fluid for recipe %s", recipe);
logger.warn("No residue specified for %s", recipe);
logger.error("Not a valid residue found for %s", recipe);
```

If everything was okay, and the recipe was valid you should see

``` java
logger.info("Adding pressing recipe %s", recipe);
```

if none of that happened, I may have been the one to pooch ;x

**Comment by PitchBright** [comment]
Just tried enabling logger in cellar.conf… tried it on both server and client. I'm not seeing any logging activity in the console, or fml-client-latest.log, or latest.log… on either the client or server in SMP, or in the client on SSP.

Should it output to console happen as I'm trying to press the fruit, or should it appear early on in the logs after startup?

**Comment by PitchBright** [comment]
I also confirmed that I had entered the Pumpkin and Reeds exactly as you've got 'em above too, so it wasn't a typo in the config on my end or anything.   :S

**Comment by PitchBright** [comment]
Turns out they were typos in the config. IceD caught my gaff's. Thanks mang. Confirmed fixed. 

Feature implemented and working perfectly. Closing Feature Request.
